zend_abstract_interface/functions/php8/functions.c

```diff
@@ -0,0 +1,139 @@
+     *
+     * https://github.com/php/php-src/blob/PHP-8.0.3/Zend/zend_execute_API.c#L673
+     */
+    ZVAL_UNDEF(retval);
+
+    if (!name || !name_len) return false;
+
+    /* Functions cannot be called outside of a request context.
+     * PG(modules_activated) indicates that all of the module RINITs have been
+     * called and we are in a request context.
+     */
+    if (!PG(modules_activated)) return false;
+
+    assert(z_is_lower(name) && "Function names must be lowercase.");
+    assert(*name != '\\' && "Function names must not have a root scope prefix '\\'.");
+
+    zend_function *func = NULL;
+    zend_string *zsname = NULL;
+    ALLOCA_FLAG(use_heap)
+
+    /* We look up the function handler directly from zend_fetch_function()
+     * instead of calling zend_is_callable_ex() because:
+     *
+     *   - We always call functions by their string name and not one of the
+     *     many 'callable' types, so the code path can be simplified.
+     *   - We assume the caller always provides the lowercased function name so
+     *     we can avoid zend_str_tolower_copy().
+     *   - We assume the caller will not prefix the function name with a
+     *     root-scope prefix, '\'.
+     *
+     * In addition, we look up the function handler from zend_fetch_function()
+     * instead of directly from the EG(function_table) to ensure that the
+     * runtime cache for userland functions will be initialized.
+     */
+    ZSTR_ALLOCA_ALLOC(zsname, name_len, use_heap);
+    memcpy(ZSTR_VAL(zsname), name, (name_len + 1));
+    func = zend_fetch_function(zsname);
+    ZSTR_ALLOCA_FREE(zsname, use_heap);
+
+    /* "function %s() does not exist" */
+    if (!func) return false;
+
+    zend_fcall_info_cache fcc = {0};
+    fcc.function_handler = func;
+
+    zend_fcall_info fci = {0};
+    fci.size = sizeof(zend_fcall_info);
+    fci.retval = retval;
+
+    /* We add the zval args directly from the variable arguments API instead of
+     * using zend_fcall_info_argv() which allows us to NULL-check each arg. It
+     * also allows us to avoid an unnecessary call to
+     * zend_fcall_info_args_clear() since the 'param{s|_count}' members in
+     * 'zend_fcall_info' will always be zeroed out in our case.
+     *
+     * Based on zend_fcall_info_argv():
+     * https://github.com/php/php-src/blob/PHP-8.0.4/Zend/zend_API.c?#L3608-L3622
+     */
+    if (argc > 0) {
+        va_list argv;
+        va_start(argv, argc);
+
+        zval *arg;
+        fci.param_count = (uint32_t)argc;
+        fci.params = (zval *)ecalloc(1, (fci.param_count * sizeof(zval)));
+
+        for (uint32_t i = 0; i < (uint32_t)argc; ++i) {
+            arg = va_arg(argv, zval *);
+            if (!arg) {
+                /* Only dtor the zvals that were successfully copied */
+                fci.param_count = i;
+                zend_fcall_info_args_clear(&fci, /* free_mem */ true);
+                return false;
+            }
+            ZVAL_COPY(&fci.params[i], arg);
```

Comment:
> why should we copy the zval to invoke a function instead of refcounting +1?

[It does both](https://github.com/php/php-src/blob/dd86987b2cbead2af1e95460cc5cebe03e83f736/Zend/zend_types.h#L1259-L1262). We need to copy the `zval` because we've allocated heap memory for the `params` on `zend_fcall_info` and we need to copy the `zval`s over to it. But after the above discussion about avoiding the heap allocation altogether, I might refactor this a bit.

Reply:
I'm not sure how that `because` is related (the only thing is, if you do not copy you cannot use zend_fcall_info_args_clear below. But that's not an issue...)? In fact zend_call_function itself already does a copy. We do not need a copy nor RC increase here at all.

Reply:
> I'm not sure how that because is related (the only thing is, if you do not copy you cannot use zend_fcall_info_args_clear below. But that's not an issue...)?

Indeed, the heap allocation for the params was left over from expanding `zend_fcall_info_argv`, so once I removed that and the `zend_fcall_info_args_clear` calls in 8ae546bf, that simplified the code quite a bit.

**Fun fact:** I originally just expanded `zend_fcall_info_argv` in place just so that I could NULL-check the args, but after this discussion, I think we're onto something even better. :)

> In fact zend_call_function itself already does a copy. We do not need a copy nor RC increase here at all.

I didn't realize that the args were being copied and having the refcount increased from within `zend_call_function`. Thanks for the tip!

I'm not sure how we would take the args from the variable arguments `va_arg` and use them in the `zend_fcall_info` params without [some kind of copy](https://github.com/DataDog/dd-trace-php/blob/8ae546bf5e42a61a3e0e0dbe475a7b1e4644253c/zend_abstract_interface/functions/php8/functions.c#L75). But I may have misunderstood what you meant by not needing to copy?

Reply:
I meant the ZVAL_COPY (vs bare assignment) - so that's fine now :-)